### PR TITLE
[FLA] Keep sigmoid(beta) in fp32 in GDN packed decode and KDA replay decode kernels

### LINKED ISSUE #38975: [Bug] GDN packed decode and KDA replay decode round sigmoid(beta) through bf16; recurrent-state drift accumulates with context length
## Describe the bug

The GDN packed decode kernel computes the beta gate as follows.

```python
# python/sglang/kernels/ops/attention/fla/fused_recurrent.py:255
beta_val = tl.sigmoid(b_val).to(b.dtype.element_ty).to(tl.float32)
```

`tl.sigmoid` is evaluated in fp32, the result is rounded to the input dtype (bf16 in serving), then cast back to fp32. No bf16 arithmetic happens between the two casts, so the round-trip only discards the low mantissa bits of the fp32 sigmoid. The rounded beta then feeds the delta-rule update.

```python
b_v *= beta_val
b_h += b_v[:, None] * b_k[None, :]
```

`b_h` is the persistent SSM state, loaded and written back on every decode step, so this rounding error enters the state once per token and accumulates with context length.

The identical line also sits in the KDA replay decode kernel at `python/sglang/kernels/ops/attention/fla/fused_recurrent_linear_replayssm.py:138`.

Two references show that keeping fp32 is the intended behavior.

1. In the same `fused_recurrent.py`, the KDA packed decode kernel keeps beta in fp32 on purpose (lines 490–493).

   ```python
   # Keep beta in fp32 (no bf16 round-trip) to match the generic decode
   # kernel `fused_sigmoid_gating_delta_rule_update`, which is the reference
   # validated against torch.
   beta_val = tl.sigmoid(b_val).to(tl.float32)
   ```

2. vllm carries a same-origin copy of this kernel (the packed decode kernel originated in vllm-project/vllm#36596; sglang#20627 imported it byte-identical, pre-fix). vllm fixed the identical line in vllm-project/vllm#53877 because "the beta rounding error accumulates with context length and can eventually degrade generation accuracy and stability". The fix was never ported here. Both sglang locations still match vllm's pre-fix code character for character (verified on current main `92dffebe`, 2026-09-10).

The suggested fix is one line per file, the same as vllm's fix.

```python
beta_val = tl.sigmoid(b_val)
```

### Measured impact

Verified on A100-PCIE-40GB, torch 2.9.0+cu129, kernel invoked directly at commit `9a1b1d2d5e`.

1. **Single step.** In the degenerate config (B=H=HV=K=V=1, zero initial state) the stored state after one decode step equals `sigmoid(b)` exactly, so the comparison reads the round-trip straight out of the state pool. State = `0.62109375` vs fp32 reference `0.622459352016449`, a rel err of 0.22%, which is exactly one bf16 rounding. With the round-trip removed, the state matches fp32 bit-exactly.

2. **1000-step decode drift** (B=1, H=16, HV=48, K=V=128, bf16 activations, fp32 state, vs an independent torch fp32 recurrence that keeps beta in fp32).

   | variant            | state rel-L2 @ step 1000 | max output rel-L2 |
   | ------------------ | ------------------------ | ----------------- |
   | HEAD               | 1.41e-3                  | 2.19e-3           |
   | round-trip removed | 1.05e-7                  | 2.39e-7           |

   The gap between the two variants is four orders of magnitude.

3. **Attribution control.** The KDA packed decode kernel in the same file already keeps beta in fp32 (point 1). Setting its per-K gate to the GDN head scalar makes the two kernels mathematically identical, so the only remaining semantic difference is whether beta is rounded through bf16. Running the two HEAD kernels head-to-head on identical inputs yields state rel-L2 1.409e-3, matching HEAD-vs-reference (1.409e-3) to three significant digits, so the entire drift is attributable to the beta round-trip, independent of any patched copy.

On the vllm side, the same line caused downstream serving regressions (flagos-ai/vllm-plugin-FL#385). sglang serves the same model families through these kernels (`qwen3_next.py`, `qwen3_5.py` via the GDN backend; the Kimi Linear family via the KDA replay path) with bf16 gating input, so the trigger conditions are identical.

The failure is silent. Outputs stay finite and well-formed, the per-step error sits inside any per-op tolerance, and an elementwise comparison against a reference will not catch it. The error only surfaces in long-recurrence comparisons or model-level evals. The vllm fix also caught a test-side pitfall worth checking here. Their old reference test computed its oracle as `torch.sigmoid(b.float()).to(dtype)`, so the reference carried the same round-trip and stayed green. Any sglang FLA reference test doing the same is blind to this bug.

## Reproduction

```python
import torch
from sglang.kernels.ops.attention.fla.fused_recurrent import (
    fused_recurrent_gated_delta_rule_packed_decode as packed_decode,
)

device, dtype = "cuda", torch.bfloat16
mixed_qkv = torch.ones((1, 3), device=device, dtype=dtype)
a = torch.zeros((1, 1), device=device, dtype=dtype)
b = torch.full((1, 1), 0.5, device=device, dtype=dtype)
params = torch.zeros((1,), device=device, dtype=dtype)
idx = torch.ones((1,), device=device, dtype=torch.int32)
state = torch.zeros((2, 1, 1, 1), device=device, dtype=torch.float32)
out = torch.empty((1, 1, 1, 1), device=device, dtype=dtype)

# Zero initial state + unit inputs: after one step the stored state equals
# sigmoid(b) itself, so the print reads the bf16 round-trip directly from
# the SSM state pool.
packed_decode(mixed_qkv=mixed_qkv, a=a, b=b, A_log=params, dt_bias=params,
              scale=1.0, initial_state=state, out=out, ssm_state_indices=idx)

got = state[1, 0, 0, 0].item()
expected = torch.sigmoid(b.float()).item()
print(f"state={got!r}  fp32 sigmoid={expected!r}  rel_err={abs(got - expected) / expected:.4%}")
# state=0.62109375  fp32 sigmoid=0.622459352016449  rel_err=0.2194%
```

This is a port of the regression test vllm added alongside its fix (`test_packed_decode_keeps_beta_in_fp32`). After changing line 255 to `beta_val = tl.sigmoid(b_val)`, the printed state matches `expected` exactly.

## Environment

- sglang commit `9a1b1d2d5e`; both lines verified still present on current main (`92dffebe`, 2026-09-10)
- NVIDIA A100-PCIE-40GB, torch 2.9.0+cu129
- Reproduction imports the kernel module directly (`sglang.kernels.ops.attention.fla.fused_recurrent`)

### PR DESCRIPTION
**Pull request by GowayLee**
## Motivation

Fixes #38975.

The GDN packed decode kernel and the KDA replay decode kernel round the fp32 sigmoid gate through the input dtype and back:

```python
# fused_recurrent.py:255, fused_recurrent_linear_replayssm.py:138
beta_val = tl.sigmoid(b_val).to(b.dtype.element_ty).to(tl.float32)
```

No bf16 arithmetic happens between the two casts, so the round-trip only discards the low mantissa bits of the fp32 sigmoid. The rounded gate feeds the delta-rule update of the persistent SSM state, which is read and written on every decode step, so the rounding error enters the state once per token and accumulates with context length.

Two references show fp32 is the intended behavior. The KDA packed decode kernel in the same file already keeps beta in fp32 on purpose (`fused_recurrent.py:490`, comment "Keep beta in fp32 (no bf16 round-trip) to match the generic decode kernel `fused_sigmoid_gating_delta_rule_update`, which is the reference validated against torch"). And the same-origin vllm copy of the packed kernel had the identical line and fixed it in vllm-project/vllm#53877 because "the beta rounding error accumulates with context length and can eventually degrade generation accuracy and stability"; that fix was never ported here.

## Modifications

One line per file, the same as vllm's fix:

```python
beta_val = tl.sigmoid(b_val)
```

- `python/sglang/kernels/ops/attention/fla/fused_recurrent.py:255` (GDN packed decode)
- `python/sglang/kernels/ops/attention/fla/fused_recurrent_linear_replayssm.py:138` (KDA replay decode)

Added `test/registered/kernel/ops/attention/test_decode_keeps_beta_in_fp32.py`, a port of the regression test added alongside the vllm fix (`test_packed_decode_keeps_beta_in_fp32`), extended to cover the replay decode kernel. Degenerate config (zero initial state, unit q=k=v), so the stored state after one decode step equals `sigmoid(b)` itself and the test reads the gate straight out of the state pool, asserted against a torch fp32 reference at rtol/atol 1e-6.

## Accuracy Tests

Verified on A100-PCIE-40GB, torch 2.9.0+cu129, at the parent commit (full data in #38975).

Single step: pre-fix state `0.62109375` vs fp32 reference `0.622459352016449` (rel err 0.2194%, exactly one bf16 rounding); post-fix bit-exact.

1000-step decode drift (B=1, H=16, HV=48, K=V=128, bf16 activations, fp32 state, vs an independent torch fp32 recurrence that keeps beta in fp32):

| variant | state rel-L2 @ step 1000 | max output rel-L2 |
| --- | --- | --- |
| pre-fix | 1.41e-3 | 2.19e-3 |
| post-fix | 1.05e-7 | 2.39e-7 |

The new unit test fails on the pre-fix kernels (both test methods) and passes post-fix. The existing `test_linear_replayssm_decode.py` kernel-consistency suite (6 tests) passes post-fix.

## Speed Tests and Profiling

The change removes two elementwise dtype casts on a scalar per program; no measurable impact is expected, so no benchmark was run.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to the [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to the [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidelines).

---
### CI States

Latest PR Test (Base): :x: [Run #34571858915](https://github.com/sgl-project/sglang/actions/runs/34571858915)
Latest PR Test (Extra): :x: [Run #34571858765](https://github.com/sgl-project/sglang/actions/runs/34571858765)
Latest PR Test (AMD ROCm 10): :x: [Run #34571858901](https://github.com/sgl-project/sglang/actions/runs/34571858901)

### CODE REVIEW

**Comment by Leslie360** [comment]
Thanks for the packed-decode fix. We hit the same defect class on the extend/prefill + NPU path (fused_gdn_gating narrows sigmoid(b) into an fp32 buffer) and opened #40362 to cover it — different files and paths, either can land without the other. Combined verification on our side: this PR + #40362 + main passes the verify-vs-decode bit-exact test (14/14), i.e. the forced-reject divergence class is closed across both paths.

One heads-up from stacking PRs while testing: #36014 as currently written moves target-verify the *other* way — it narrows verify beta to bf16 to match the pre-this-PR packed decode. With this PR merged, packed decode keeps fp32, so #36014 would re-open the exact gap it set out to close (our combo test went 0/128 matching with all three applied). If #36014 is revived, its verify path would need to keep fp32 instead.
